fs/span-manager/span_manager.go

```diff
@@ -173,104 +193,46 @@ func (m *SpanManager) buildAllSpans() {
 	}
 }
 
-// fetchWithRetries fetches the requested data and verifies that the span digest matches the one in the ztoc.
-// It will retry the fetch and verification m.maxSpanVerificationFailureRetries times.
-// It does not retry when there is an error fetching the data, because retries already happen lower in the stack in httpFetcher.
-// If there is an error fetching data from remote, it is not an transient error.
-func (m *SpanManager) fetchWithRetries(spanID compression.SpanID, buf []byte, offset compression.Offset) error {
-	var (
-		err error
-		n   int
-	)
-	for i := 0; i < m.maxSpanVerificationFailureRetries+1; i++ {
-		n, err = m.r.ReadAt(buf, int64(offset))
-		if err != nil {
-			return err
-		}
-
-		if n != len(buf) {
-			return fmt.Errorf("unexpected data size for reading compressed span. read = %d, expected = %d", n, len(buf))
-		}
-
-		if err = m.verifySpanContents(buf, spanID); err != nil {
-			continue
-		}
-		return nil
-	}
-	return err
-}
-
-// FetchSingleSpan invokes the reader to fetch the span in the background and cache it.
-// It is invoked by the BackgroundFetcher.
+// FetchSingleSpan invokes the reader to fetch the span in the background and cache
+// the span without uncompressing. It is invoked by the BackgroundFetcher.
+// span state change: unrequested -> requested -> fetched.
 func (m *SpanManager) FetchSingleSpan(spanID compression.SpanID) error {
 	if spanID > m.ztoc.CompressionInfo.MaxSpanID {
 		return ErrExceedMaxSpan
 	}
 
+	// return directly if span is not in `unrequested`
 	s := m.spans[spanID]
-	s.mu.Lock()
-	state := s.state.Load().(spanState)
-
-	// Only fetch if the span hasn't been requested yet.
-	if state != unrequested {
-		s.mu.Unlock()
+	if !s.checkState(unrequested) {
 		return nil
 	}
 
-	s.setState(requested)
-	s.mu.Unlock()
-
-	compressedSize := s.endCompOffset - s.startCompOffset
-	compressedBuf := make([]byte, compressedSize)
-
-	if err := m.fetchWithRetries(spanID, compressedBuf, s.startCompOffset); err != nil {
-		return err
-	}
-
 	s.mu.Lock()
 	defer s.mu.Unlock()
-
-	state = s.state.Load().(spanState)
-	if state != requested {
+	// check again after acquiring Lock
+	if !s.checkState(unrequested) {
 		return nil
 	}
 
-	m.addSpanToCache(fmt.Sprintf("%d", spanID), compressedBuf)
-	return s.setState(fetched)
+	_, err := m.fetchAndCacheSpan(spanID, false)
+	return err
 }
 
+// ResolveSpan ensures a span exists uncompressed in cache by calling
+// `resolveSpan`. (no span data is explicitly requested from `ResolveSpan`).
 func (m *SpanManager) ResolveSpan(spanID compression.SpanID) error {
```

Comment:
This existing`ResolveSpan` is only used for testing. I didn't delete it but I don't think it's useful.

Reply:
I'm not a fan of an exported function that's only used for testing.

Reply:
updated to private func